treq/auth.py

```diff
@@ -1,7 +1,144 @@
+                   - nonce - "nonce" HTTP Digest authentication param
+                   - qop - Quality Of Protection HTTP Digest auth param
+                   - opaque - "opaque" HTTP Digest authentication param
+                         (should be sent back to server unchanged)
+                   - cached - Identifies that authentication already have been
+                         performed for URI/method,
+                         and new request should use the same params as first
+                         authenticated request
+                   - path - the URI path where we are authenticating
+                   - method - HTTP method to be used when requesting
+    :return: HTTP Digest authentication string
+    """
+    algo = kwargs.get('algorithm', 'MD5').upper()
+    original_algo = kwargs.get('algorithm')
+    qop = kwargs.get('qop', None)
+    nonce = kwargs.get('nonce')
+    opaque = kwargs.get('opaque', None)
+    path_parsed = urlparse(kwargs['path'])
+    actual_path = path_parsed.path
+
+    if path_parsed.query:
+        actual_path += '?' + path_parsed.query
+
+    a1 = '%s:%s:%s' % (
+        agent.username,
+        kwargs['realm'],
+        agent.password
+    )
+
+    a2 = '%s:%s' % (
+        kwargs['method'],
+        actual_path
+    )
+
+    if algo == 'MD5' or algo == 'MD5-SESS':
+        digest_hash_func = _md5_utf_digest
+    elif algo == 'SHA':
+        digest_hash_func = _sha1_utf_digest
+    else:
+        raise UnknownDigestAuthAlgorithm(algo)
+
+    ha1 = digest_hash_func(a1)
+    ha2 = digest_hash_func(a2)
+
+    cnonce = generate_client_nonce(nonce)
+
+    if algo == 'MD5-SESS':
+        ha1 = digest_hash_func("%s:%s:%s" % (ha1, nonce, cnonce), algo)
+
+    if kwargs['cached']:
+        agent.digest_auth_cache[(kwargs['method'], kwargs['path'])]['c'] += 1
+        nonce_count = agent.digest_auth_cache[
+            (kwargs['method'], kwargs['path'])
+        ]['c']
+    else:
+        nonce_count = 1
+
+    ncvalue = '%08x' % nonce_count
+    if qop is None:
+        response_digest = digest_hash_func(
+            "%s:%s" % (ha1, "%s:%s" % (ha2, nonce))
+        )
+    else:
+        noncebit = "%s:%s:%s:%s:%s" % (
+            nonce, ncvalue, cnonce.encode('utf-8'), 'auth', ha2
+        )
+        response_digest = digest_hash_func("%s:%s" % (ha1, noncebit))
+
+    hb = 'username="%s", realm="%s", nonce="%s", uri="%s", response="%s"' % (
+        agent.username, kwargs['realm'], nonce, actual_path, response_digest
+    )
+    if opaque:
+        hb += ', opaque="%s"' % opaque
+    if original_algo:
+        hb += ', algorithm="%s"' % original_algo
+    if qop:
+        hb += ', qop="auth", nc=%s, cnonce="%s"' % (ncvalue, cnonce)
+    if not kwargs['cached']:
+        agent.digest_auth_cache[(kwargs['method'], kwargs['path'])] = {
+            'p': kwargs,
+            'c': 1
+        }
+    return 'Digest %s' % hb
+
+
+class HTTPDigestAuth(object):
```

Comment:
This should probably just use https://attrs.readthedocs.io/en/stable/

(see also: https://glyph.twistedmatrix.com/2016/08/attrs.html)